libazureinit/src/provision/user.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::process::Command;
+
+use tracing::instrument;
+
+use crate::error::Error;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Provisioner {
+    Useradd,
+    #[cfg(test)]
+    FakeUseradd,
+}
+
+impl Provisioner {
+    pub(crate) fn create(self, name: impl AsRef<str>) -> Result<(), Error> {
+        match self {
+            Self::Useradd => useradd(name.as_ref()),
+            #[cfg(test)]
+            Self::FakeUseradd => Ok(()),
+        }
+    }
+}
+
+#[instrument(skip_all)]
+fn useradd(name: &str) -> Result<(), Error> {
+    let path_useradd = env!("PATH_USERADD");
+    let home_path = format!("/home/{name}");
+
+    let status = Command::new(path_useradd)
+                    .arg(name)
+                    .arg("--comment")
+                    .arg(
+                      "Provisioning agent created this user based on username provided in IMDS",
+                    )
+                    .arg("--groups")
+                    .arg("adm,audio,cdrom,dialout,dip,floppy,lxd,netdev,plugdev,sudo,video")
```

Comment:
For Azure Linux, we will want to specify an alternative list of default groups for the user ([example](https://github.com/Azure/azure-init/pull/87/files#diff-be2f6d40c6bbf8fc0882a218ec29f5579d7b1d83c4449d9efe6814c64b6ba3e3R42)).

There's a few different ways to specify this, but perhaps an environment variable is the most logical? As that is how we are handling distro-specific knowledge in the other situations (i.e., binary paths).



To illustrate current behavior, here is a test trying to call path_useradd with this `--groups` list, and there are a few groups that aren't recognized on AzL. (See photo)
<img width="317" alt="image" src="https://github.com/Azure/azure-init/assets/22079170/7a3a3c0b-c8da-4d0f-af7b-68c4a43dd2e7">
_*Note: this test was built off of #100 so there may be some additional differences in the code, but the result of calling `useradd` should be the same as at the time of this writing both PRs use the [same ](https://github.com/Azure/azure-init/blob/5171131d549c29db7d465368dc9d407024b27830/libazureinit/src/distro.rs#L19)--groups list_

Reply:
Hm, yes. I added a commit to this PR that sets it via environment variable, but I'll mull on it more over the weekend.

I do worry about the discoverability of this approach. I think it's a little weird to have a library accept configuration this way instead of exposing an API and letting the application using the library handle it (either via build-time environment variables, or CLI options, or what have you).

Reply:
Okay, I added a `User` type in commit 3b4b94b583b1d4e92b96a6a5d0d0c7b1345b859a and threaded it through the API. It bundles all the user-related configuration together and allows you to optionally provide a list of supplementary groups. How does that look to you?